msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenByAuthorizationGrantSupplier.java

```diff
@@ -77,55 +77,56 @@ private boolean IsUiRequiredCacheSupported() {
     private OAuthAuthorizationGrant processPasswordGrant(
             OAuthAuthorizationGrant authGrant) throws Exception {
 
-        if (!(authGrant.getAuthorizationGrant() instanceof ResourceOwnerPasswordCredentialsGrant)) {
+        if (!(authGrant.toParameters().get(GrantConstants.GRANT_TYPE_PARAMETER).get(0).equals(GrantConstants.PASSWORD))) {
             return authGrant;
         }
 
         if (msalRequest.application().authenticationAuthority.authorityType != AuthorityType.AAD) {
             return authGrant;
         }
 
-        ResourceOwnerPasswordCredentialsGrant grant =
-                (ResourceOwnerPasswordCredentialsGrant) authGrant.getAuthorizationGrant();
-
         UserDiscoveryResponse userDiscoveryResponse = UserDiscoveryRequest.execute(
-                this.clientApplication.authenticationAuthority.getUserRealmEndpoint(grant.getUsername()),
+                this.clientApplication.authenticationAuthority.getUserRealmEndpoint(authGrant.toParameters().get("username").get(0)),
                 msalRequest.headers().getReadonlyHeaderMap(),
                 msalRequest.requestContext(),
                 this.clientApplication.serviceBundle());
 
         if (userDiscoveryResponse.isAccountFederated()) {
             WSTrustResponse response = WSTrustRequest.execute(
                     userDiscoveryResponse.federationMetadataUrl(),
-                    grant.getUsername(),
-                    grant.getPassword().getValue(),
+                    authGrant.toParameters().get(GrantConstants.USERNAME_PARAMETER).get(0),
+                    authGrant.toParameters().get(GrantConstants.PASSWORD_PARAMETER).get(0),
                     userDiscoveryResponse.cloudAudienceUrn(),
                     msalRequest.requestContext(),
                     this.clientApplication.serviceBundle(),
                     this.clientApplication.logPii());
 
-            AuthorizationGrant updatedGrant = getSAMLAuthorizationGrant(response);
+            Map<String, List<String>> params = getSAMLAuthGrantParameters(response);
+            params.putAll(authGrant.toParameters());
 
-            authGrant = new OAuthAuthorizationGrant(updatedGrant, authGrant.getParameters());
+            authGrant = new OAuthAuthorizationGrant(params, null, null);
         }
         return authGrant;
     }
 
-    private AuthorizationGrant getSAMLAuthorizationGrant(WSTrustResponse response) throws UnsupportedEncodingException {
-        AuthorizationGrant updatedGrant;
+    private Map<String, List<String>> getSAMLAuthGrantParameters(WSTrustResponse response) {
```

Comment:
Now maybe I see why you went with the "expose a map everywhere" strategy. Maybe OAuthGrantParameters is too specific a case, and we need a more generic "paramter bag" type?

Reply:
Yeah, pretty much every nimbus API used `Map<String, List<String>>` for parameters. The next PR will replace the final util imports that still use that map format and use something simpler, this PR was trying to just do a 1:1 replacement to lower the risk of behavioral changes.